rust/src/applayertemplate/mod.rs

```diff
@@ -0,0 +1,518 @@
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&TemplateTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self) -> TemplateTransaction {
+        let mut tx = TemplateTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn find_request(&mut self) -> Option<&mut TemplateTransaction> {
+        for tx in &mut self.transactions {
+            if tx.response.is_none() {
+                return Some(tx);
+            }
+        }
+        None
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> bool {
+        // We're not interested in empty requests.
+        if input.len() == 0 {
+            return true;
+        }
+
+        // For simplicity, always extend the buffer and work on it.
+        self.request_buffer.extend(input);
+
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.request_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            match parser::parse_message(current) {
+                nom::IResult::Done(rem, request) => {
+                    current = rem;
+
+                    println!("Request: {}", request);
+                    let mut tx = self.new_tx();
+                    tx.request = Some(request);
+                    self.transactions.push(tx);
+                }
+                nom::IResult::Incomplete(_) => {
+                    self.request_buffer.extend_from_slice(current);
+                    break;
+                }
+                nom::IResult::Error(_) => {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    fn parse_response(&mut self, input: &[u8]) -> bool {
+        // We're not interested in empty responses.
+        if input.len() == 0 {
+            return true;
+        }
+
+        // For simplicity, always extend the buffer and work on it.
+        self.response_buffer.extend(input);
+
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.response_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            match parser::parse_message(current) {
+                nom::IResult::Done(rem, response) => {
+                    current = rem;
+
+                    match self.find_request() {
+                        Some(tx) => {
+                            tx.response = Some(response);
+                            println!("Found response for request:");
```

Comment:
should we use SCLog* here?

Reply:
Thought about it? But I had every intention that the user would remove that stuff there and not convert to SCLogDebug.  I'm indifferent about it though.

Reply:
I prefer SCLog, probably SCLogNotice in the template that the dev can change to Debug